### ISSUE TITLE
mountSuspended with i18n

### ISSUE THREAD

**Issue by ronaldkonjer** [nuxt/test-utils#585]
I'm setting up the nuxt-vitest module in my project. Though I'm running into some issues getting a simple test, I use mountSuspended(Logo) to test the mounting of a Logo component. This component uses i18n. Whereas when I'm using mount from Vitest the test passes.

The error I'm getting is as follows:
```
Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Rejection ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
SyntaxError: Need to install with `app.use` function
 ❯ Module.createCompileError ../node_modules/.pnpm/@intlify+message-compiler@9.3.0-beta.17/node_modules/@intlify/message-compiler/dist/message-compiler.mjs:54:19
 ❯ createI18nError ../node_modules/.pnpm/vue-i18n@9.3.0-beta.17_vue@3.3.4/node_modules/vue-i18n/dist/vue-i18n.runtime.mjs:97:34
     96|     [I18nErrorCodes.INVALID_ARGUMENT]: 'Invalid argument',
     97|     [I18nErrorCodes.MUST_BE_CALL_SETUP_TOP]: 'Must be called at the top of a `setup` function',
     98|     [I18nErrorCodes.NOT_INSLALLED]: 'Need to install with `app.use` function',
       |                                  ^
     99|     [I18nErrorCodes.UNEXPECTED_ERROR]: 'Unexpected error',
    100|     [I18nErrorCodes.NOT_AVAILABLE_IN_LEGACY_MODE]: 'Not available in legacy mode',
 ❯ Module.useI18n ../node_modules/.pnpm/vue-i18n@9.3.0-beta.17_vue@3.3.4/node_modules/vue-i18n/dist/vue-i18n.runtime.mjs:2253:15
 ❯ setup app.vue:33:40
```

And while we are at it. Is there a way to test with the translated i18n messages iso the keys?

######

Here are some snippets from my configuration.
vitest.config.js
```
...
export default defineVitestConfig({
  resolve: {
    alias
  },
  test: {
    dir: tests,
    setupFiles: ['tests/unit.setup.ts'],
    environment: 'jsdom',
    globals: true,
    root: rootDir,
    environmentOptions: {
      nuxt: {
        rootDir
      }
    }
  }
})
```

./test/nuxt/components/Logo.nuxt.spec.ts (with mountSuspended)
```
import { describe, it, expect } from 'vitest'

import { mountSuspended } from 'vitest-environment-nuxt/utils'
import Logo from '@/components/header/Logo.vue'

describe('HeaderLogo', () => {
  it('can mount some component', async () => {
    const component = await mountSuspended(Logo)
    expect(component.vm).toBeTruthy()
    expect(component.text()).toMatchInlineSnapshot(
      '"global.logo-brand-name.global.logo-brand-localehomepage.logo-tagline"'
    )
  })
})
```
./tests/unit.setup.ts
```
import { config } from '@vue/test-utils'
import { createI18n } from 'vue-i18n'

import nlNL from '~/locales/nl_NL.json'

const i18n = createI18n({
  legacy: false,
  locale: 'nl_NL',
  missing: (_, key) => key,
  messages: {
    nlNL
  }
})

config.global.mocks = {
  t: msg => msg
}
config.global.plugins.push(i18n)
```

./tests/nuxt/components/Logo.nuxt.spec.ts (without mountSuspended)
```
import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'

import Logo from '@/components/header/Logo.vue'

describe('HeaderLogo', () => {
  it('can mount some component', () => {
    const component = mount(Logo)
    expect(component.vm).toBeTruthy()
    expect(component.text()).toMatchInlineSnapshot(
      '"global.logo-brand-name.global.logo-brand-localehomepage.logo-tagline"'
    )
  })
})
```

**Comment by danielroe** [comment]
If someone is still experiencing this, would you provide a reproduction?

I've opened a sample PR here with an runtime test with i18n which seems to be working fine: https://github.com/nuxt/test-utils/pull/633

**Comment by sarayourfriend** [comment]
@danielroe I have a reproduction in a public repository, WordPress/openverse, where we're attemping to migrate to Nuxt 3 and are running into this very problem.

Unfortunately things are a bit messy there still, because we're also still working through other issues: https://github.com/WordPress/openverse/pull/3571

You can reproduce the issue by running `pnpm run test:unit v-image-cell` inside the `frontend` directory.

What I'm seeing when I run vitest in the `frontend` directory is that `globalThis.useNuxtApp().$i18n` has a working vuei18n instance. I can pass it keys and I get back the right strings. However, if I retrieve `vueApp` from `globalThis.__unctx__.get("nuxt-app").vueApp` and then access `vueApp._context.provides[vueApp.__VUE_I18N_SYMBOL__]`, that vuei18n object has zero configured instances, it doesn't look installed. Though, I'm not sure I'm looking at this right. I'm not familiar with the guts of vue-i18n and have just been digging through the test-utils code to figure out how this module works to set up the render function's context with the configured Nuxt environment.

If you'd like me to open a separate issue for this, please let me know. The exact same presentation for us as in this issue:

```
vue:setup
  [nuxt-app] vue:setup: 259.213ms []
vue:error
  [nuxt-app] vue:error: 0.816ms [
    SyntaxError: Need to install with `app.use` function
        at Module.createCompileError (/var/home/sara/projects/openverse/node_modules/.pnpm/@intlify+message-compiler@9.8.0/node_modules/@intlify/message-compiler/dist/message-compiler.mjs:78:19)
        at createI18nError (/var/home/sara/projects/openverse/node_modules/.pnpm/vue-i18n@9.8.0_vue@3.3.12/node_modules/vue-i18n/dist/vue-i18n.mjs:105:34)
        at Module.useI18n (/var/home/sara/projects/openverse/node_modules/.pnpm/vue-i18n@9.8.0_vue@3.3.12/node_modules/vue-i18n/dist/vue-i18n.mjs:2313:15)
        at setup (/var/home/sara/projects/openverse/frontend/src/components/VImageCell/VImageCell.vue:138:11)
        at clonedComponent.setup (/var/home/sara/projects/openverse/node_modules/.pnpm/@nuxt+test-utils@3.9.0_@testing-library+vue@8.0.1_@vitest+ui@1.1.0_@vue+test-utils@2.4.3_h3@1_prr4umkikytza6o2cai2dcbbye/node_modules/@nuxt/test-utils/dist/runtime-utils/index.mjs:215:37)
        at callWithErrorHandling (/var/home/sara/projects/openverse/node_modules/.pnpm/@vue+runtime-core@3.3.12/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:156:18)
        at setupStatefulComponent (/var/home/sara/projects/openverse/node_modules/.pnpm/@vue+runtime-core@3.3.12/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7285:25)
        at setupComponent (/var/home/sara/projects/openverse/node_modules/.pnpm/@vue+runtime-core@3.3.12/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7246:36)
        at mountComponent (/var/home/sara/projects/openverse/node_modules/.pnpm/@vue+runtime-core@3.3.12/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:5643:7)
        at processComponent (/var/home/sara/projects/openverse/node_modules/.pnpm/@vue+runtime-core@3.3.12/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:5609:9) {
      code: 27,
      domain: undefined
    },
    {
      image: [Getter/Setter],
      searchTerm: [Getter/Setter],
      aspectRatio: [Getter/Setter],
      kind: [Getter/Setter],
      relatedTo: [Getter/Setter]
    },
    'setup function'
  ]
page:finish
  [nuxt-app] page:finish: 261.044ms []
```

**Comment by sarayourfriend** [comment]
Actually, I was able to reproduce this inside the application: https://github.com/nuxt/test-utils/pull/673

The issue is with the `useI18n` composable. Using the template (which the example didn't test either) via `$t` OR using `useNuxtApp().$i18n` _does_ work. So it seems like there's some kind of vue-i18n wiring that is getting lost in the Nuxt environment construction.

**Comment by renky** [comment]
Same here, but it seems like setting globals to true solves it for me

I'm using composition api with setup script, and i use the translation already in this setup-function:

```
<script setup lang="ts">
...
     const { t } = useI18n()
...
    some function usint t('....') something
...
```

and when testing this Component, i exactly get the same error like above:
```
SyntaxError: Need to install with `app.use` function
 ❯ Module.createCompileError node_modules/@intlify/message-compiler/dist/message-compiler.mjs:78:19
     76|         : code;
     77|     const error = new SyntaxError(String(msg));
     78|     error.code = code;
       |                   ^
     79|     if (loc) {
     80|         error.location = loc;
 ❯ createI18nError node_modules/vue-i18n/dist/vue-i18n.mjs:105:34
 ❯ Module.useI18n node_modules/vue-i18n/dist/vue-i18n.mjs:2314:15
 ❯ setup components/Form/Form.vue:39:11
 ❯ wrappedSetup node_modules/@nuxt/test-utils/dist/runtime-utils/index.mjs:90:26
 ❯ clonedComponent.setup node_modules/@nuxt/test-utils/dist/runtime-utils/index.mjs:138:48
 ❯ callWithErrorHandling node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:193:18
 ❯ setupStatefulComponent node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7546:25
 ❯ setupComponent node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7507:36
 ❯ mountComponent node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:5836:7
```


But finally, when I switch on globals in vitest config, the error is gone:
vitest.config.ts:
```
export default defineVitestConfig({
  test: {
    globals: true
  }
})
```

**Comment by renky** [comment]
Can confirm @sarayourfriend
It's just happening when using 
```
const { t } = useI18n()
```

after changing all my t()-function-calls to 
```
useNuxtApp().$i18n.t()
```
and removing the above useI18n-composable call, the error is gone.

**Comment by peterleilei86** [comment]
@renky It is indeed gone but `useNuxtApp().$i18n().t()` cannot be used if you want component based translation. It can only be done using `useI18n` like 
```js
const { t } = useI18n({useScope: 'local', messages: { en: { test: 'test' } } } })
```

`i18n` component option does not work either for local scoped translations!

**Comment by martio** [comment]
Spent two days on this. It's worth fixing this issue. I solved the problem by mocking the "vue-i18n" module and assigning it to global plugins

`tests/setup.ts`

```javascript
import { beforeAll, afterAll, vi } from 'vitest'
import { ref } from 'vue'
import { config } from '@vue/test-utils'
import { createI18n } from 'vue-i18n'

beforeAll(() => {
  vi.mock('vue-i18n', async (importOriginal) => {
    const module = await importOriginal<typeof import('vue-i18n')>()

    return {
      ...module,
      useI18n: () => ({
        locale: ref('en'),
        locales: ref([
          {
            name: 'English',
            code: 'en',
            iso: 'en-GB',
            file: 'en.ts',
          },
        ]),
      }),
    }
  })
})

afterAll(() => {
  vi.restoreAllMocks()
})

// setup i18n
const i18n = createI18n({
  legacy: false,
  locale: 'en',
  locales: [
    {
      name: 'English',
      code: 'en',
      iso: 'en-GB',
      file: 'en.ts',
    },
  ],
  messages: {},
  missing: (locale: string, key: string) => key,
})

config.global.plugins.push(i18n)
```

**Comment by maciekiwaniuk** [comment]
@martio solution work for me, thank you

**Comment by sniperadmin** [comment]
@martio Solution is perfect :)

**Comment by rigtigeEmil** [comment]
> Spent two days on this. It's worth fixing this issue. I solved the problem by mocking the "vue-i18n" module and assigning it to global plugins
> 
> `tests/setup.ts`
> 
> import { beforeAll, afterAll, vi } from 'vitest'
> import { ref } from 'vue'
> import { config } from '@vue/test-utils'
> import { createI18n } from 'vue-i18n'
> 
> beforeAll(() => {
>   vi.mock('vue-i18n', async (importOriginal) => {
>     const module = await importOriginal<typeof import('vue-i18n')>()
> 
>     return {
>       ...module,
>       useI18n: () => ({
>         locale: ref('en'),
>         locales: ref([
>           {
>             name: 'English',
>             code: 'en',
>             iso: 'en-GB',
>             file: 'en.ts',
>           },
>         ]),
>       }),
>     }
>   })
> })
> 
> afterAll(() => {
>   vi.restoreAllMocks()
> })
> 
> // setup i18n
> const i18n = createI18n({
>   legacy: false,
>   locale: 'en',
>   locales: [
>     {
>       name: 'English',
>       code: 'en',
>       iso: 'en-GB',
>       file: 'en.ts',
>     },
>   ],
>   messages: {},
>   missing: (locale: string, key: string) => key,
> })
> 
> config.global.plugins.push(i18n)

This does not work for me, is there something missing before this is doable (or am I just missing an obvious step)?
`[Vue warn]: Directive "t" has already been registered in target app.`

**Comment by gbyesiltas** [comment]
> Spent two days on this. It's worth fixing this issue. I solved the problem by mocking the "vue-i18n" module and assigning it to global plugins
> 
> `tests/setup.ts`
> 
> import { beforeAll, afterAll, vi } from 'vitest'
> import { ref } from 'vue'
> import { config } from '@vue/test-utils'
> import { createI18n } from 'vue-i18n'
> 
> beforeAll(() => {
>   vi.mock('vue-i18n', async (importOriginal) => {
>     const module = await importOriginal<typeof import('vue-i18n')>()
> 
>     return {
>       ...module,
>       useI18n: () => ({
>         locale: ref('en'),
>         locales: ref([
>           {
>             name: 'English',
>             code: 'en',
>             iso: 'en-GB',
>             file: 'en.ts',
>           },
>         ]),
>       }),
>     }
>   })
> })
> 
> afterAll(() => {
>   vi.restoreAllMocks()
> })
> 
> // setup i18n
> const i18n = createI18n({
>   legacy: false,
>   locale: 'en',
>   locales: [
>     {
>       name: 'English',
>       code: 'en',
>       iso: 'en-GB',
>       file: 'en.ts',
>     },
>   ],
>   messages: {},
>   missing: (locale: string, key: string) => key,
> })
> 
> config.global.plugins.push(i18n)

I'm not quite sure why but this didn't work for me. Instead, I was able to make it working with:

```
// vitest.setup.ts

mockNuxtImport("useI18n", async () => {
  const actualNuxtAppImports =
    await vi.importActual<typeof import("#app")>("#app");

  return () => {
    return actualNuxtAppImports.useNuxtApp().$i18n;
  };
});
```

**Comment by shovv** [comment]
Thank you @martio, dzięki 🥇

**Comment by redlizard** [comment]
@danielroe Can we reopen this issue? There has been a reproduction since shortly after the bot closed the issue, which I think you have investigated, but the ticket has remained closed.

**Comment by danielroe** [comment]
@redlizard would you mind opening a new issue with a minimal reproduction? there's a lot here that's not relevant any more

**Comment by redlizard** [comment]
I don't think there's anything irrelevant in this issue. There's the good reproduction in #673 , and basically everything else in this thread is reasonable workarounds to the problem. But I can create a new issue if you prefer.

**Comment by gbyesiltas** [comment]
@danielroe @redlizard Not to conflate issues, but I suspect [this issue](https://github.com/nuxt/test-utils/issues/1152)￼ is related. Many people work around it by manually registering the plugin in vitest.setup, which results in the plugin being registered twice. If the double-instance problem is real and gets fixed, this may stop being an issue altogether, since the plugin already appears to be registered once by default.

**Comment by redlizard** [comment]
I suspect the same. All my efforts to debug this issue or work around it have been consistent with a duplicate Vue instance being created by `mountSuspended` and the different instances then being used in different places.

As an example, I first noticed the problem in this issue when I tried to register i18n in vitest.setup (or in mountSuspended()), as you say, and then found that useNuxtApp().$i18n.setLocale() doesn't update useI18n().locale, which sure sounds like there are multiple vue instances in play.

**Comment by gbyesiltas** [comment]
> [@danielroe](https://github.com/danielroe) [@redlizard](https://github.com/redlizard) Not to conflate issues, but I suspect [this issue](https://github.com/nuxt/test-utils/issues/1152)￼ is related. Many people work around it by manually registering the plugin in vitest.setup, which results in the plugin being registered twice. If the double-instance problem is real and gets fixed, this may stop being an issue altogether, since the plugin already appears to be registered once by default.

okay, coming back to my own comment, it actually looks like there was already a pr merged for the issue that I linked and if I update the nuxt/test-utils to the latest version, the i18n module also just works without any workarounds :) I'm gonna close that original issue

**Comment by redlizard** [comment]
It appears this issue is also fixed, by #1477 . That `patchInstanceAppContext` correctly transfers the vue-i18n context over to the `mountSuspended` vue instance.

I missed it because I was on the latest nuxt-3 (3.20) and matching nuxt/test-utils, while the above fix landed in 3.21. I had not realized nuxt and nuxt/test-utils diverged version-wise.

Thanks for the hard work everyone, looks like this one is squished.
